Frameworks/UIKit/NSTextContainer.mm

```diff
@@ -62,38 +78,121 @@ - (NSArray*)exclusionPaths {
     return _exclusionPaths;
 }
 
+// Finds the first viable region for the proposed rect given the exclusion areas based upon x positions
+// Assumes writing direction is left to right, exclusionLines is sorted and preprocessed for y positions, and all inputs are valid
+static CGRect __FirstPossibleRectForProposed(CGRect proposed,
+                                             const std::vector<CGRect>& exclusionLines,
+                                             CGFloat padding,
+                                             CGFloat maxWidth) {
+    for (auto path : exclusionLines) {
+        if (proposed.origin.x < path.origin.x) {
+            proposed.size.width = path.origin.x - proposed.origin.x - padding;
+            if (proposed.size.width > padding) {
+                break;
+            }
+        }
+        if (path.origin.x + path.size.width > proposed.origin.x) {
+            proposed.origin.x = (path.origin.x + path.size.width) + padding;
+            proposed.size.width = maxWidth - proposed.origin.x;
+        }
+    }
+
+    return proposed;
+}
+
+// Finds the first point at which the given rect intersects the given exclusionContext, moving by delta each iteration
```

Comment:
>moving by delta each iteration [](start = 88, length = 30)

is there a more efficient way to do this? this current implementation seems incredibly expensive.